http://bugs.gentoo.org/137799
http://gcc.gnu.org/PR28058

2006-08-27  Mark Mitchell  <mark@codesourcery.com>

	PR c++/28058
	* pt.c (register_specialization): Return error_mark_node for
	specialization-after-instantiation.
	* decl2.c (mark_used): Mark the main function used when one of its
	clones is used.

--- branches/gcc-4_1-branch/gcc/cp/pt.c	2006/08/27 17:45:25	116491
+++ branches/gcc-4_1-branch/gcc/cp/pt.c	2006/08/27 17:59:55	116492
@@ -1177,7 +1177,7 @@
 	    {
 	      error ("specialization of %qD after instantiation",
 		     fn);
-	      return spec;
+	      return error_mark_node;
 	    }
 	  else
 	    {
--- branches/gcc-4_1-branch/gcc/cp/decl2.c	2006/08/27 17:45:25	116491
+++ branches/gcc-4_1-branch/gcc/cp/decl2.c	2006/08/27 17:59:55	116492
@@ -3256,6 +3256,8 @@
     }
 
   TREE_USED (decl) = 1;
+  if (DECL_CLONED_FUNCTION_P (decl))
+    TREE_USED (DECL_CLONED_FUNCTION (decl)) = 1;
   /* If we don't need a value, then we don't need to synthesize DECL.  */ 
   if (skip_evaluation)
     return;
